model.py

```diff
@@ -166,6 +191,58 @@ class SoftplusLayer(tf.keras.layers.Layer):
     def call(self, inputs):
         return tf.math.log(1 + tf.exp(inputs))
 
+def cross_val_with_best_model(
+    build_model_fn: Callable[[], Model],
+    evaluate_fn: Callable[[Model, pd.DataFrame, pd.DataFrame], float],
+    sp: ScaledPartitions,
+    n_cv_folds: int,
+    cache_best_model=True,
+    output_model_filename: str=None,
+    **fit_kwargs):
+    '''
+        Runs cross validation on the model, returning the history and fitted best model. 
+        build_model_fn: Function that builds, compiles, and returns a model.
+        evaluate_fn: Function for scoring and comparing models across different folds.
+        sp: Stands for "Scaled Partitions", representing your dataset. This function only looks at the "train" 
+            partition and uses this to create validation sets.
+        n_cv_folds: Number of cross validation folds. These folds are sources from the "train" partition.
+        evaluate_fn
+        cache_best_model: Whether or not to retain a copy of the best model in memory. May be highly inefficient
+            depending on the model.
+
+    '''
+    assert cache_best_model or output_model_filename 
+
+    kf = KFold(n_splits=n_cv_folds)
+    best_model_filepath = None
+    best_model = None
+    best_score = -float('inf')  # Initialize with a very low score
+
+    for fold, (train_index, val_index) in enumerate(kf.split(sp.train.X)):
+        print(sp.train.X.index)
+        X_train, X_val = sp.train.X.loc[train_index], sp.train.X.loc[val_index]
+        Y_train, Y_val = sp.train.Y.loc[train_index], sp.train.Y.loc[val_index]
+
+        model = build_model_fn()
+        training_artifacts = model.fit(
+            X_train=X_train, Y_train=Y_train,
+            validation_data=(X_val, Y_val), 
+            **fit_kwargs)
+
+        # Evaluate the model on the validation set
+        score = evaluate_fn(model, X_val, Y_val)
+
+        if score > best_score:
+            best_score = score
+            best_model_artifacts = training_artifacts
+            if cache_best_model:
+                best_model = deepcopy(model)
+            if output_model_filename:
+                model.save_model(output_model_filename) 
+
+    return training_artifacts, best_model
```

Comment:
It's perfectly valid (and will produce better results) to return a model trained on the entire train+validation set after performing cross-validation. Could we do this here?